File: src/main/java/com/example/visainterview/service/AccountService.java
package com.example.visainterview.service;

import com.example.visainterview.dto.AccountResponse;
import com.example.visainterview.dto.CreateAccountRequest;
import com.example.visainterview.entity.Account;
import com.example.visainterview.entity.Customer;
import com.example.visainterview.exception.AccountNotFoundException;
import com.example.visainterview.exception.DuplicateDocumentNumberException;
import com.example.visainterview.repository.AccountRepository;
import com.example.visainterview.repository.CustomerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;

@Service
@RequiredArgsConstructor
public class AccountService {

        private final AccountRepository accountRepository;
        private final CustomerRepository customerRepository;

        @Transactional
        public AccountResponse createAccount(CreateAccountRequest request) {
                if (customerRepository.existsByDocumentNumber(request.getDocumentNumber())) {
                        throw new DuplicateDocumentNumberException("Document number already exists");
                }

                Customer customer = Customer.builder()
                                .documentNumber(request.getDocumentNumber())
                                .build();
                customer = customerRepository.save(customer);

                Account account = Account.builder()
                                .customer(customer)
                                .balance(BigDecimal.ZERO)
                                .build();
                account = accountRepository.save(account);

                return AccountResponse.builder()
                                .accountId(account.getId())
                                .documentNumber(customer.getDocumentNumber())
                                .balance(account.getBalance())
                                .build();
        }

        @Transactional(readOnly = true)
        public AccountResponse getAccount(Long accountId) {
                Account account = accountRepository.findById(accountId)
                                .orElseThrow(() -> new AccountNotFoundException("Account not found"));

                return AccountResponse.builder()
                                .accountId(account.getId())
                                .balance(account.getBalance())
                                .documentNumber(account.getCustomer().getDocumentNumber())
                                .build();
        }
}

File: golang/service/account_service.go
package service

import (
	"context"
	"errors"

	"github.com/visainterview/financial-transaction-system/dto"
	"github.com/visainterview/financial-transaction-system/middleware"
	"github.com/visainterview/financial-transaction-system/models"
	"github.com/visainterview/financial-transaction-system/repository"
	"github.com/visainterview/financial-transaction-system/utils"
	"gorm.io/gorm"
)

type AccountService interface {
	CreateAccount(ctx context.Context, req dto.CreateAccountRequest) (*dto.AccountResponse, error)
	GetAccount(ctx context.Context, accountID uint) (*dto.AccountResponse, error)
}

type accountService struct {
	accountRepo repository.AccountRepository
}

func NewAccountService(accountRepo repository.AccountRepository) AccountService {
	return &accountService{accountRepo: accountRepo}
}

func (s *accountService) CreateAccount(ctx context.Context, req dto.CreateAccountRequest) (*dto.AccountResponse, error) {
	if req.DocumentNumber == "" {
		return nil, utils.ErrDocumentNumberRequired
	}

	existing, err := s.accountRepo.FindByDocumentNumber(ctx, req.DocumentNumber)
	if err == nil && existing != nil {
		return nil, utils.ErrDocumentNumberExists
	}

	customer := &models.Customer{
		DocumentNumber: req.DocumentNumber,
	}

	account := &models.Account{
		Balance: 0,
		Version: 0,
	}

	if err := s.accountRepo.CreateCustomerAndAccount(ctx, customer, account); err != nil {
		return nil, err
	}

	middleware.RecordAccountCreation()

	return &dto.AccountResponse{
		AccountID:      account.ID,
		DocumentNumber: customer.DocumentNumber,
		Balance:        account.Balance,
	}, nil
}

func (s *accountService) GetAccount(ctx context.Context, accountID uint) (*dto.AccountResponse, error) {
	account, err := s.accountRepo.FindByID(ctx, accountID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, utils.ErrAccountNotFound
		}
		return nil, err
	}

	return &dto.AccountResponse{
		AccountID:      account.ID,
		DocumentNumber: account.Customer.DocumentNumber,
		Balance:        account.Balance,
	}, nil
}
